### ISSUE TITLE
CMake: install target not working: fails: file INSTALL cannot find absl-config.cmake

### ISSUE THREAD

**Issue by phillip-keldenich** [google/or-tools#1025]
I tried building and installing the source code from git (I tried both master and tags/v7.0-beta1 with the same error) on Mac OS, installing as dependencies only wget/cmake (using homebrew).

Configuring, building and even testing using the standard CMake workflow

```
mkdir build && cd build
cmake ..
make -j8
```

 work fine (except for tons of warnings regarding the deprecated "register" keyword and some .cc-files without symbols).

However, running make install (or, equivalently, cmake --build . --target install) fails because the file build/abseil-cpp-src/CMake/absl-config.cmake is missing.
I tried to run 
```
find build | grep absl-config
```
too see whether the file is present somewhere else, but there appears to be no such file in the build directory.

```
-- Installing: /usr/local/lib/cmake/protobuf/protobuf-module.cmake
-- Installing: /usr/local/lib/cmake/protobuf/protobuf-options.cmake
-- Installing: /usr/local/lib/cmake/protobuf/protobuf-config.cmake
-- Installing: /usr/local/lib/cmake/protobuf/protobuf-config-version.cmake
-- Installing: /usr/local/lib/cmake/absl/absl-targets.cmake
-- Installing: /usr/local/lib/cmake/absl/absl-targets-release.cmake
CMake Error at abseil-cpp-build/cmake_install.cmake:55 (file):
  file INSTALL cannot find
  "/Users/me/Source/or-tools/build/abseil-cpp-src/CMake/absl-config.cmake".
Call Stack (most recent call first):
  external/cmake_install.cmake:41 (include)
  cmake_install.cmake:103 (include)
```

If I manually add an empty file in its place, the install goes through without error. However, as expected, I cannot use ortools in my cmake project with an empty absl-config.cmake file installed:
```
find_package(ortools CONFIG REQUIRED)
```
yields the following error:
```
CMake Error at CMakeLists.txt:12 (find_package):
  Found package configuration file:

    /usr/local/lib/cmake/ortools/ortoolsConfig.cmake

  but it set ortools_FOUND to FALSE so package "ortools" is considered to be
  NOT FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  absl::absl_base absl::absl_container absl::absl_hash absl::absl_memory
  absl::absl_meta absl::absl_str_format absl::absl_strings
  absl::absl_synchronization absl::absl_types
```
I then proceeded to find absl-config.cmake in /patches (in the source directory).
This file seems to contain a syntax error at least in tags/v7.0-beta1 (missing closing parenthesis).
I guess something was supposed to copy that file to the build directory where install expects it; I did that manually, retried the install, and it worked.

I still had no luck consuming ortools cmake config files in my cmake project, still with the same error message as before.
Adding a
```find_package(absl CONFIG REQUIRED)```
before that fixed this issue, so that configuring, building and linking (in my own CMake project) against ortools::ortools now works.

Was I doing something wrong here, like missing a "patching" step when running cmake? Then that should maybe be included somewhere in the documentation.

**Comment by Mizux** [comment]
hi,

Thanks for the feedback, I need to rework the abseil-cpp integration against https://github.com/abseil/abseil-cpp/pull/182 .

or-tools should depends on `absl::base` not `absl::absl_base`

**Comment by lihuiknight** [comment]
@phillip-keldenich I also try to install v7.0-beta1 based on source code, and meet the following error:
./ortools/algorithms/knapsack_solver.h:65:32: fatal error: absl/memory/memory.h: No such file or directory

Do you fix this error?

**Comment by phillip-keldenich** [comment]
No, for me, building (compiling etc.) worked fine, only installation didn't. This error sounds like some missing include directories somewhere in the build process. I also did not try to install ortools using pip (i.e., python package management), but directly using the source code.

I guess that fixing any issues will be harder for you than for me since you are not (completely) in control of the build process and debugging a build done on-the-fly by a package manager like pip will at least be quite inconvenient. Without further information, it also is not clear whether the problem is a bug in the pip package (possible) or ortools itself (also possible).

**Comment by Mizux** [comment]
The main issue with CMake based build, is you can't have install rule if you integrate your dependencies using the [`add_subdirectory()/ExternalProject()` hack](https://crascit.com/2015/07/25/cmake-gtest/).

TLDR: when you install a cmake target (e.g. ortools) you must also install all `target_link_libraries()` targets which are not **IMPORTED** target (e.g. absl::base)
bonus limitation: [you can't install target ALIAS](https://cmake.org/cmake/help/v3.13/command/add_library.html#alias-libraries) (say hello to absl::base target **ALIAS** :smile: , note: [yes you can retrieve the original target name](https://github.com/google/or-tools/pull/1019/files#diff-ed5192e5b0d17843cb9549861c9e2516R58))
bonus limitation: before cmake 3.13 [you can't install target define outside of the current *directory* scope](https://cmake.org/cmake/help/v3.13/release/3.13.html#commands).

Todo(mizux): create a repository and longer/detailed explanation why it doesn't scale.

note: PR #1019 will help a little, but the wall process of dependencies must be rework to use an installed dependencies based CMake  build.
note: you can find a nice topic on this subject (install dependencies) when abseil-cpp was refusing to have install target https://github.com/abseil/abseil-cpp/issues/38

**Comment by Mizux** [comment]
Don't want to block the release v7.0 which BTW only use the Makefile-based build to generate all its binary artifacts.
So I will postpone it to 7.1 since we need a deep rework of all the CMake (experimental) stuff.
